### PR TITLE
Add reduce_view

### PR DESCRIPTION
**Pull request by fzimmermann89**
Utility for reducing expanded/broadcasted dimensions in a view to singleton

### CODE REVIEW

**Comment by fzimmermann89** [comment]
(open to name or docstring changes)

**Comment by fzimmermann89** [comment]
@ckolbPTB @schuenke 
can you review this? Or should I close it and try to get it into the indexing pr-stack? 
I thought initially, it could be easier without using ghstack, but I lose track which commits I need to get stuff to work...
So this PR and #484 are logically kind-of parts of the stack starting at #488...

File: src/mrpro/utils/reshape.py [hunk 1]
```diff
@@ -67,3 +69,32 @@ def broadcast_right(*x: torch.Tensor) -> tuple[torch.Tensor, ...]:
     max_dim = max(el.ndim for el in x)
     unsqueezed = torch.broadcast_tensors(*(unsqueeze_right(el, max_dim - el.ndim) for el in x))
     return unsqueezed
+
+
+def reduce_view(x: torch.Tensor, dim: int | Sequence[int] | None = None) -> torch.Tensor:
+    """Replace expanded dimensions in a view by singletons.
```
Comment (ckolbPTB) [inline src/mrpro/utils/reshape.py]:
```suggestion
    """Reduce expanded dimensions in a view to singletons.
```

File: src/mrpro/utils/reshape.py [hunk 2]
```diff
@@ -67,3 +69,32 @@ def broadcast_right(*x: torch.Tensor) -> tuple[torch.Tensor, ...]:
...
+    Reduce either all or a specific dimension to a singleton only if it
+    points to the same memory address.
+    This undoes expand.
+
+    Parameters
+    ----------
+    x
+        input tensor
+    dim
+        only undo expands in the specified dimensions.
```
Comment (ckolbPTB) [inline src/mrpro/utils/reshape.py]:
```suggestion
        only reduce expanded dimensions in the specified dimensions.
```

File: src/mrpro/utils/reshape.py [hunk 3]
```diff
@@ -67,3 +69,32 @@ def broadcast_right(*x: torch.Tensor) -> tuple[torch.Tensor, ...]:
     max_dim = max(el.ndim for el in x)
     unsqueezed = torch.broadcast_tensors(*(unsqueeze_right(el, max_dim - el.ndim) for el in x))
     return unsqueezed
+
+
+def reduce_view(x: torch.Tensor, dim: int | Sequence[int] | None = None) -> torch.Tensor:
+    """Replace expanded dimensions in a view by singletons.
+
+    Reduce either all or a specific dimension to a singleton only if it
```
Comment (ckolbPTB) [inline src/mrpro/utils/reshape.py]:
```suggestion
    Reduce either all or specific dimensions to a singleton if it
```

File: src/mrpro/utils/reshape.py [hunk 4]
```diff
@@ -67,3 +69,32 @@ def broadcast_right(*x: torch.Tensor) -> tuple[torch.Tensor, ...]:
...
+    points to the same memory address.
+    This undoes expand.
+
+    Parameters
+    ----------
+    x
+        input tensor
+    dim
+        only undo expands in the specified dimensions.
+        If None, undo expand in all dimensions.
```
Comment (ckolbPTB) [inline src/mrpro/utils/reshape.py]:
```suggestion
        If None, reduce all expanded dimensions.
```

File: src/mrpro/utils/reshape.py [hunk 5]
```diff
@@ -67,3 +69,32 @@ def broadcast_right(*x: torch.Tensor) -> tuple[torch.Tensor, ...]:
...
+    """
+    if dim is None:
+        dim_: Sequence[int] = range(x.ndim)
+    elif isinstance(dim, Sequence):
+        dim_ = [d % x.ndim for d in dim]
+    else:
+        dim_ = [dim % x.ndim]
+
+    stride = x.stride()
```
Comment (ckolbPTB) [inline src/mrpro/utils/reshape.py:96]: can we avoid using stride and size inside and outside of the expression?

**Review by ckolbPTB — APPROVED** [review]

**Pull request merged**